Работа с внешними API

JuicyFlash opened the pull request:
- Используется официальный клиент GitHub Octokit
- Выводится ссылка на созданный gist во flash сообщении
- Добавлена локализация в сервис GistQuestionService
- Добавлена модель\миграция\ресурс Gist
- Администратору добавлен вывод таблицы всех созданных gist. Таблица содержит колонки:
     1)Вопрос. Отображаются первые 25 символов тела вопроса в виде ссылки. По клике на ссылку отображается страница вопроса (действие show)
     2)Gist URL. Отображается хэш созданного gist в виде ссылки
     3)Автор. Отображается email пользователя который создал gist

- Токен помещён в .env файл, используется dotenv gem

JuicyFlash committed:
- Используется официальный клиент GitHub Octokit
- Выводится ссылка на созданный gist во flash сообщении
- Добавлена локализация в сервис GistQuestionService
- Добавлена модель\миграция\ресурс Gist
- Администратору добавлен вывод таблицы всех созданных gist. Таблица содержит колонки:
     1)Вопрос. Отображаются первые 25 символов тела вопроса в виде ссылки. По клике на ссылку отображается страница вопроса (действие show)
     2)Gist URL. Отображается хэш созданного gist в виде ссылки
     3)Автор. Отображается email пользователя который создал gist

- Токен помещён в .env файл, используется dotenv gem

BubuntuClu commented on app/models/gist.rb:
```diff
@@ -0,0 +1,15 @@
+  belongs_to :author, class_name: "User" , foreign_key: :author_id
+
+  def gist_question_title
+    question.body.slice(0..24)
```
можешь использоватье метод `truncate` для обрезания текста

BubuntuClu commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,49 @@
+  def initialize(question, client: nil)
+    @question = question
+    @test = @question.test
+    @client = client || Octokit::Client.new(:access_token => ENV['ACCESS_TOKEN'])
```
замени передачу параметров без использования хэшрокета

BubuntuClu commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,49 @@
+
+  def call
+    client.create_gist(gist_params)
+    self
```
а зачем такой костыль? почему бы просто не инициализировать сервис отдельно от вызова метода `call`?

BubuntuClu commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,49 @@
+  def success?
+    if client.last_response.nil?
+      false
+    elsif client.last_response.status == 201
```
что вернет тебе метод, если статус будет 200/300/404? любой отличный от 201

------------
elsif это плохая конструкция

```
return false if client.last_response.nil?

client.last_response.status == 201
```
можно заменить на такой код, но его тоже надо доработать

BubuntuClu commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,49 @@
+  end
+
+  def gist_url
+    client.last_response.data.html_url unless client.last_response.nil?
```
старайся всегда исопльзовать `if` вместо `unless`, так как `if` воспринимается легче

BubuntuClu commented on db/migrate/20220526182943_create_gists.rb:
```diff
@@ -0,0 +1,10 @@
+      t.string :url
+      t.timestamps
+    end
+    add_reference(:gists, :question, foreign_key: { to_table: :questions })
```
почему эти две оперции выполнены вне блока `create_table`?
поиести их внутрь, чтобы создавалась таблица за один запрос, а не за 3

BubuntuClu commented on app/controllers/processed_tests_controller.rb:
```diff
@@ -20,6 +20,19 @@ def update
+
+      if result.success?
+         flash_options = { notice: t('.gist_success', url: result.gist_url)  }
+         current_user.author_gists.new(question:@processed_test.current_question, url:result.gist_url).save
```
`new(....).save` равносильно `create(...)`
почему ты предпочел первый вариант использовать?

JuicyFlash committed:
применён truncate для обрезания текста
заменил хэшрокет
инициализация отдельно от метода GistQuestionService.call
изменил метод GistQuestionService.success? (как можно его доработать ещё не придумал...)
создание связей таблицы перенёс в блок create_table
применён меток create вместо save

BubuntuClu commented on db/migrate/20220526182943_create_gists.rb:
```diff
@@ -0,0 +1,10 @@
+      t.string :url
+      t.timestamps
+      t.references :author, foreign_key: { to_table: :users }
+      t.references :question, foreign_key: { to_table: :questions }
```
тут не надо указывать таблицу. имя связи совпадает с именем таблицы, так что можно просто `foreign_key: true`
